packages/protocol-http/src/httpHandler.ts

```diff
@@ -3,4 +3,19 @@ import { HttpHandlerOptions, RequestHandler } from "@smithy/types";
 import { HttpRequest } from "./httpRequest";
 import { HttpResponse } from "./httpResponse";
 
-export type HttpHandler = RequestHandler<HttpRequest, HttpResponse, HttpHandlerOptions>;
+/**
+ * @internal
+ */
+export type HttpHandler<HttpHandlerConfig> = RequestHandler<HttpRequest, HttpResponse, HttpHandlerOptions> & {
+  /**
+   * @internal
+   * @param key
+   * @param value
+   */
+  updateHttpClientConfig(key: keyof HttpHandlerConfig, value: HttpHandlerConfig[typeof key]): void;
```

Comment:
I think there is too much type magic going on here.
Can we not just assume it can be any generic key / value pair and remove the generic?

Reply:
I don't think this is magic? It is saying that if you call updateConfig, it should be for a key and value type that is known to exist on the config's type.

If it feels restrictive, a compromise would be to intersect `HttpHandlerConfig` with `{ [key: string]: unknown }` for unbounded additional values. But, I don't think that should be necessary. 

For real type shenanigans, you should see what I committed to https://github.com/awslabs/smithy-typescript/blob/main/packages/types/src/transform/type-transform.ts